### PR TITLE
gateway: fix etag and set xattr before complete multipart rename to ensure consistency

### PR DESCRIPTION
**Pull request by zhijian-pro**

### CODE REVIEW

File: pkg/gateway/gateway.go [hunk 1]
```diff
@@ -1075,6 +1082,25 @@ func (n *jfsObjects) CompleteMultipartUpload(ctx context.Context, bucket, object
...
+		eno = n.fs.SetXattr(mctx, tmp, s3Etag, []byte(s3MD5), 0)
+		if eno != 0 {
+			logger.Warnf("set xattr error, path: %s,xattr: %s,value: %s,flags: %d", tmp, s3Etag, s3MD5, 0)
+		}
+	}
+
+	var tagStr []byte
+	if n.gConf.ObjTag {
+		var eno syscall.Errno
+		if tagStr, eno = n.fs.GetXattr(mctx, n.upath(bucket, uploadID), s3Tags); eno != 0 && eno != meta.ENOATTR {
```
Comment (davies) [inline pkg/gateway/gateway.go:1097]:
```
		if tagStr, eno = n.fs.GetXattr(mctx, n.upath(bucket, uploadID), s3Tags); eno != 0  {
                        if  eno != meta.ENOATTR {
   			logger.Errorf("get object tags error, path: %s, error: %s", n.upath(bucket, uploadID), eno)
                        }
		} else if eno = n.fs.SetXattr(mctx, tmp, s3Tags, tagStr, 0); eno != 0 {
			logger.Errorf("set object tags error, path: %s, tags: %s, error: %s", tmp, string(tagStr), eno)
		}
```

File: pkg/gateway/gateway.go [hunk 2]
```diff
@@ -793,7 +790,7 @@ func (n *jfsObjects) PutObject(ctx context.Context, bucket string, object string
 		return
 	}
 	var tagStr string
-	var etag string
+	var etag = r.MD5CurrentHexString()
```
Comment (davies) [inline pkg/gateway/gateway.go:793]: this could be wrong (before read)

File: pkg/gateway/gateway.go [hunk 3]
```diff
@@ -802,30 +807,31 @@ func (n *jfsObjects) PutObject(ctx context.Context, bucket string, object string
...
+	} else {
+		if err = n.putObject(ctx, bucket, p, r, opts, func(tmpName string) {
+			etag = r.MD5CurrentHexString()
+			if n.gConf.KeepEtag && !strings.HasSuffix(object, sep) {
+				if eno := n.fs.SetXattr(mctx, tmpName, s3Etag, []byte(etag), 0); eno != 0 {
+					logger.Errorf("set xattr error, path: %s,xattr: %s,value: %s,flags: %d", tmpName, s3Etag, etag, 0)
+				}
+			}
+			// tags: key1=value1&key2=value2&key3=value3
+			var tagStr string
```
Comment (jiefenghuang) [inline pkg/gateway/gateway.go:819]: overwrite outside var?

**Review by jiefenghuang — APPROVED** [review]
LGTM, maybe add more unit tests covering gateway cases

**Pull request merged**